File: dokkaJavaApi/src/main/kotlin/com/virtuslab/dokka/site/processors.kt
package dokka.java.api.com.virtuslab.dokka.site

import org.jetbrains.dokka.DokkaConfiguration
import org.jetbrains.dokka.base.renderers.html.NavigationNode
import org.jetbrains.dokka.base.renderers.html.NavigationPage
import org.jetbrains.dokka.links.DRI
import org.jetbrains.dokka.model.Documentable
import org.jetbrains.dokka.model.properties.PropertyContainer
import org.jetbrains.dokka.pages.*
import org.jetbrains.dokka.plugability.DokkaContext
import org.jetbrains.dokka.transformers.pages.PageTransformer
import java.io.File

const val ExternalDocsTooKey = "ExternalDocsTooKey"

val docsRootDRI = DRI.topLevel.copy(extra = "_top_level_index")

data class PreRenderedContent(
    val html: String,
    override val dci: DCI,
    override val sourceSets: Set<DokkaConfiguration.DokkaSourceSet>,
    override val style: Set<Style> = emptySet(),
    override val extra: PropertyContainer<ContentNode> = PropertyContainer.empty()
) : ContentNode {
    override fun hasAnyContent(): Boolean = !html.isBlank()

    override fun withNewExtras(newExtras: PropertyContainer<ContentNode>): ContentNode = copy(extra = newExtras)
}

abstract class BaseStaticSiteProcessor(cxt: DokkaContext) : PageTransformer {

    final override fun invoke(input: RootPageNode): RootPageNode =
        rawRoot?.let { transform(input) } ?: input

    protected abstract fun transform(input: RootPageNode): RootPageNode

    private val rawRoot: File? = cxt.configuration.pluginsConfiguration.get(ExternalDocsTooKey)?.let { File(it) }
    val root = rawRoot ?: File("unknown")

    protected val mySoruceSet = cxt.configuration.sourceSets.toSet()
    protected val docsFile = File(root, "docs")

    protected fun File.asDri(): DRI {
        val relativePath = root.toPath().relativize(toPath()).toString().replace(File.separatorChar, '.')
        return DRI("_.$relativePath")
    }

    inner class DocPageNode(
        private val from: File,
        private val template: TemplateFile?,
        override val children: List<DocPageNode>,
        val resolved: ResolvedPage,
        override val dri: Set<DRI>,
        override val embeddedResources: List<String> = emptyList()
    ) : ContentPage {
        override val name: String = template?.name() ?: from.name

        val isIndexPage = from.name == "index.md" || from.name == "index.html"

        fun title() = template?.title() ?: name

        override val documentable: Documentable? = null
        override val content: ContentNode = PreRenderedContent(resolved.html, DCI(dri, ContentKind.Empty), mySoruceSet)

        override fun modified(
            name: String,
            content: ContentNode,
            dri: Set<DRI>,
            embeddedResources: List<String>,
            children: List<PageNode>
        ): ContentPage =
            DocPageNode(from, template, children.filterIsInstance<DocPageNode>(), resolved, dri, embeddedResources)

        override fun modified(name: String, children: List<PageNode>): PageNode =
            DocPageNode(from, template, children.filterIsInstance<DocPageNode>(), resolved, dri, embeddedResources)
    }
}

class SiteResourceManager(cxt: DokkaContext) : BaseStaticSiteProcessor(cxt) {
    private fun listResources(nodes: List<PageNode>): Set<String> =
        nodes.flatMap {
            when {
                it is DocPageNode ->
                    listResources(it.children) + it.resolved.resources
                else -> emptySet()
            }
        }.toSet()


    override fun transform(input: RootPageNode): RootPageNode {
        val images = File(root, "images").walkTopDown().filter { it.isFile }.map{ root.toPath().relativize(it.toPath()).toString() }
        val resources = listResources(input.children) + images
        val resourcePages = resources.map { path ->
            RendererSpecificResourcePage(path, emptyList(), RenderingStrategy.Write(File(root, path).readText()))
        }
        val modified = input.transformContentPagesTree {
            when (it) {
                is DocPageNode -> it.modified(embeddedResources = it.embeddedResources + it.resolved.resources)
                else -> it
            }
        }
        return modified.modified(children = resourcePages + modified.children)
    }
}

class SitePagesCreator(cxt: DokkaContext) : BaseStaticSiteProcessor(cxt) {

    override fun transform(input: RootPageNode): RootPageNode {
        val (navigationPage, rest) = input.children.partition { it is NavigationPage }
        val defaultNavigation = (navigationPage.single() as NavigationPage).root

        val (indexes, children) = loadFiles().partition { it.isIndexPage }
        if (indexes.size > 1) println("ERROR: Multiple index pages found ${children.filter { it.isIndexPage }}") // TODO proper error handling

        fun toNavigationNode(c: DocPageNode): NavigationNode =
            NavigationNode(
                c.title(),
                c.dri.first(),
                mySoruceSet,
                c.children.map { toNavigationNode(it) }
            )

        val apiPageDri = DRI.topLevel.copy(extra = "_api_")

        val mergedRoots = NavigationNode(
            defaultNavigation.name,
            defaultNavigation.dri,
            defaultNavigation.sourceSets,
            children.map { toNavigationNode(it) } + listOf(
                NavigationNode(
                    "API",
                    apiPageDri,
                    defaultNavigation.sourceSets,
                    defaultNavigation.children
                )
            )
        )
        val original = indexes.firstOrNull()?.let { indexPage ->
            rest.map {
                when (it) {
                    is ModulePageNode ->
                        if (it.dri.contains(DRI.topLevel)) {
                            val packageList =
                                PackagePageNode(
                                    "all packages",
                                    it.content,
                                    setOf(apiPageDri),
                                    null,
                                    emptyList(),
                                    it.embeddedResources
                                )
                            it.modified(content = indexPage.content, children = it.children + packageList)
                        } else it
                    else ->
                        it
                }
            }
        } ?: rest

        val indexFiles = listOf(File(root, "index.html"), File(root, "index.md")).filter { it.exists() }
        if (indexFiles.size > 1) println("ERROR: Multiple root index pages found: ${indexFiles.map { it.absolutePath }}") // TODO proper error handling

        val topLevelIndexPage = indexFiles.take(1)
            .mapNotNull { renderDocs(it, noChildren = true)?.modified(dri = setOf(docsRootDRI)) }

        return input.modified(children = original + topLevelIndexPage + listOf(NavigationPage(mergedRoots)) + children)
    }

    private val layouts: Map<String, TemplateFile> by lazy {
        val layoutRoot = File(root, "_layouts")
        val dirs: Array<File> = layoutRoot.listFiles() ?: emptyArray()
        dirs.map { loadTemplateFile(it) }.map { it.name() to it }.toMap()
    }

    private fun renderDocs(from: File, noChildren: Boolean = false): DocPageNode? =
        if (from.name.startsWith("_")) null else try {
            val dri = setOf(from.asDri())

            val children =
                if (noChildren) emptyList() else from.listFiles()?.mapNotNull { renderDocs(it) } ?: emptyList()
            if (children.count { it.isIndexPage } > 1)
                println("ERROR: Multiple index pages found ${children.filter { it.isIndexPage }}") // TODO proper error handling

            val templateFile = if (from.isDirectory) null else loadTemplateFile(from)
            val content = try {
                val context = RenderingContext(emptyMap(), layouts)
                if (from.isDirectory) {
                    children.find { it.isIndexPage }?.resolved ?: EmptyResolvedPage
                } else templateFile!!.resolve(context)
            } catch (e: Throwable) {
                val msg = "Error rendering $from: ${e.message}"
                println("ERROR: $msg") // TODO proper error handling
                ResolvedPage(msg, emptyList())
            }

            DocPageNode(from, templateFile, children.filter { !it.isIndexPage }, content, dri)
        } catch (e: RuntimeException) {
            e.printStackTrace()
            null
        }

    private fun loadFiles(): List<DocPageNode> =
        docsFile.listFiles()?.mapNotNull { renderDocs(it) } ?: emptyList()

}

File: scala3doc/src/dotty/dokka/site/processors.scala
package dotty.dokka
package site

import java.io.File
import java.nio.file.Files
import java.nio.file.FileVisitOption

import org.jetbrains.dokka.base.renderers.html.{NavigationNode, NavigationPage}
import org.jetbrains.dokka.model.Documentable
import org.jetbrains.dokka.pages._
import org.jetbrains.dokka.transformers.pages.PageTransformer

import scala.collection.JavaConverters._

abstract class BaseStaticSiteProcessor(staticSiteContext: Option[StaticSiteContext]) extends PageTransformer:
  final override def invoke(input: RootPageNode): RootPageNode = staticSiteContext.fold(input)(transform(input, _))

  protected def transform(input: RootPageNode, ctx: StaticSiteContext): RootPageNode

class SiteResourceManager(ctx: Option[StaticSiteContext]) extends BaseStaticSiteProcessor(ctx):
  private def listResources(nodes: Seq[PageNode]): Set[String] =
    nodes.flatMap {
      case it: StaticPageNode => listResources(it.getChildren.asScala.toList) ++ it.resources()
      case _ => Seq.empty
    }.toSet

  override def transform(input: RootPageNode, ctx: StaticSiteContext): RootPageNode =
    val rootPath = ctx.root.toPath
    val imgPath = rootPath.resolve("images")
    val images =
      if !Files.exists(imgPath) then Nil
      else
        val allPaths = Files.walk(imgPath, FileVisitOption.FOLLOW_LINKS)
        val files = allPaths.filter(Files.isRegularFile(_)).iterator().asScala
        files.map(p => rootPath.relativize(p).toString).toList

    val resources = images ++ listResources(input.getChildren.asScala.toList)
    val resourcePages = resources.map { path =>
      val strategy = new RenderingStrategy.Copy(rootPath.resolve(path).toString)
      new RendererSpecificResourcePage(path, JList(), strategy)
    }

    val modified = input.transformContentPagesTree {
      case it: StaticPageNode =>
        it.copy(getEmbeddedResources =
          if it.template.hasFrame then it.getEmbeddedResources ++ it.resources().asJava
          else it.resources().asJava
        )
      case it => it
    }
    modified.modified(modified.getName, (resourcePages ++ modified.getChildren.asScala).asJava)

case class AContentPage(
  override val getName: String,
  override val getChildren: JList[PageNode],
  override val getContent: ContentNode,
  override val getDri: JSet[DRI],
  override val getEmbeddedResources: JList[String] = JList(),
) extends ContentPage:
  override def getDocumentable: Documentable = null

  override def modified(
    name: String,
    content: ContentNode,
    dri: JSet[DRI],
    embeddedResources: JList[String],
    children: JList[_ <: PageNode]
  ): ContentPage = copy(name, children.asInstanceOf[JList[PageNode]], content, dri, embeddedResources)

  override def modified(name: String, children: JList[_ <: PageNode]): PageNode =
    copy(name, getChildren = children.asInstanceOf[JList[PageNode]])

class SitePagesCreator(ctx: Option[StaticSiteContext]) extends BaseStaticSiteProcessor(ctx):
  private def processRootPage(input: RootPageNode, children: List[PageNode] = Nil): AContentPage = input match
    case input: ContentPage =>
      AContentPage(
        input.getName,
        children.asJava,
        input.getContent,
        JSet(apiPageDRI),
        input.getEmbeddedResources
      )
    case _: RendererSpecificRootPage =>
      children.filter(_.isInstanceOf[RootPageNode]) match
          case List(nestedRoot: RootPageNode) =>
            processRootPage(nestedRoot, children.filter { _ != nestedRoot } ++ nestedRoot.getChildren.asScala)
          case other =>
            throw new RuntimeException(s"Expected single nested roor but get: $other")

    case _ => throw new RuntimeException(s"UNSUPPORTED! ${input.getClass.getName}")

  override def transform(input: RootPageNode, ctx: StaticSiteContext): RootPageNode =
    val (contentPage, others) = input.getChildren.asScala.toList.partition { _.isInstanceOf[ContentPage] }
    val modifiedModuleRoot = processRootPage(input, contentPage)
    val (indexes, children) = ctx.allPages.partition(f =>
      f.template.isIndexPage() && f.template.file.toPath.getParent() == ctx.docsPath )
    // TODO (https://github.com/lampepfl/scala3doc/issues/238): provide proper error handling
    if (indexes.size > 1) println(s"ERROR: Multiple index pages found ${indexes.map(_.template.file)}")

    val rootContent = indexes.headOption.fold(ctx.asContent(Text(), mkDRI(extra = "root_content")).get(0))(_.getContent)

    val root = AContentPage(
      ctx.args.projectTitle.getOrElse(ctx.args.name),
      (List(modifiedModuleRoot.modified("API", modifiedModuleRoot.getChildren)) ++ children).asJava,
      rootContent,
      JSet(docsDRI),
      JList()
    )

    new RendererSpecificRootPage(
      modifiedModuleRoot.getName,
      (List(root) ++ others).asJava,
      RenderingStrategy.DoNothing.INSTANCE
    )

class RootIndexPageCreator(ctx: Option[StaticSiteContext]) extends BaseStaticSiteProcessor(ctx):
  override def transform(input: RootPageNode, ctx: StaticSiteContext): RootPageNode =
    ctx.indexPage().fold(input){ it =>
      val (contentNodes, nonContent) = input.getChildren.asScala.partition { _.isInstanceOf[ContentNode] }
      val (navigations, rest) = nonContent.partition { _.isInstanceOf[NavigationPage] }
      val modifiedNavigation = navigations.map { it =>
        val root = it.asInstanceOf[NavigationPage].getRoot
        val api = root.getChildren.asScala.filter(_.getDri == apiPageDRI)

        def toNavigationNode(page: StaticPageNode): NavigationNode = NavigationNode(
            page.title(),
            page.getDri.asScala.head,
            root.getSourceSets,
            page.getChildren.asScala.collect { case p: StaticPageNode => toNavigationNode(p)}.asJava
          )

        new NavigationPage(
          new NavigationNode(
            input.getName,
            docsRootDRI,
            root.getSourceSets,
            (ctx.mainPages.map(toNavigationNode) ++ api).asJava
          )
        )
      }
      val newRoot = it.copy(getDri =  JSet(docsRootDRI), getChildren = contentNodes.asJava)
      input.modified(input.getName, (List(newRoot) ++ rest ++ modifiedNavigation).asJava)
    }
